File: src/lib/db/achievements.ts
/**
 * Achievement System TypeScript Bindings
 * Connects to Rust backend via Tauri IPC
 * Falls back to in-memory storage when Tauri is unavailable
 */

import type {
  Achievement,
  AchievementProgress,
  AchievementUnlock,
  AchievementAnalytics,
  AchievementRarity,
  AchievementCategory,
} from "../../types/achievements";

// Safely import invoke, handle case where Tauri is not available
let invoke: ((command: string, args?: any) => Promise<any>) | null = null;
let isTauriAvailable = false;

try {
  const tauriModule = require("@tauri-apps/api/core");
  invoke = tauriModule.invoke;
  isTauriAvailable = true;
} catch (e) {
  console.debug("Tauri not available, using in-memory achievement storage");
  invoke = null;
}

// Re-export types for convenience
export type {
  Achievement,
  AchievementProgress,
  AchievementUnlock,
  AchievementAnalytics,
  AchievementRarity,
  AchievementCategory,
};

// ============================================
// FALLBACK IN-MEMORY STORAGE
// ============================================

// When Tauri is not available, we use in-memory storage
const fallbackStorage = {
  achievements: new Map<string, Achievement>(),
  playerStats: new Map<string, PlayerStats>(),
  progress: new Map<string, AchievementProgress[]>(),
  unlocks: new Map<string, AchievementUnlock[]>(),
};

// ============================================
// REQUEST/RESPONSE TYPES
// ============================================

interface InitAchievementsRequest {
  achievements_json: string;
}

interface UpdateProgressRequest {
  achievement_id: string;
  progress: number;
  player_id?: string;
}

interface UnlockAchievementRequest {
  achievement_id: string;
  player_id?: string;
  context_json?: string;
}

interface AddXpRequest {
  amount: number;
  player_id?: string;
}

export interface PlayerStats {
  player_id: string;
  level: number;
  xp: number;
  total_xp_earned: number;
  achievements_unlocked: number;
  current_streak: number;
  longest_streak: number;
  last_login?: number;
  created_at: number;
  updated_at: number;
}

// ============================================
// HELPER FUNCTION FOR SAFE INVOKE
// ============================================

async function safeInvoke<T>(command: string, args?: any): Promise<T> {
  if (!invoke) {
    throw new Error(`Cannot execute Tauri command "${command}" - Tauri is not available`);
  }
  return (invoke as (command: string, args?: any) => Promise<T>)(command, args);
}

// ============================================
// ACHIEVEMENT DATABASE API
// ============================================

export class AchievementDatabaseAPI {
  /**
   * Initialize achievements in the database
   * Should be called once on app startup with all achievement definitions
   */
  async initAchievements(achievements: Achievement[]): Promise<string> {
    // If Tauri is not available, use fallback
    if (!invoke) {
      for (const achievement of achievements) {
        fallbackStorage.achievements.set(achievement.id, achievement);
      }
      return "OK";
    }

    const request: InitAchievementsRequest = {
      achievements_json: JSON.stringify(achievements),
    };
    return await safeInvoke<string>("init_achievements", { request });
  }

  /**
   * Get all achievements from database
   */
  async getAllAchievements(): Promise<Achievement[]> {
    // If Tauri is not available, use fallback
    if (!invoke) {
      return Array.from(fallbackStorage.achievements.values());
    }

    const responseJson = await safeInvoke<string>("get_all_achievements");
    return JSON.parse(responseJson);
  }

  /**
   * Get specific achievement by ID
   */
  async getAchievementById(achievementId: string): Promise<Achievement | null> {
    // If Tauri is not available, use fallback
    if (!invoke) {
      return fallbackStorage.achievements.get(achievementId) || null;
    }

    const responseJson = await safeInvoke<string>("get_achievement_by_id", {
      achievementId,
    });
    return JSON.parse(responseJson);
  }

  /**
   * Get all progress records for a player
   */
  async getPlayerProgress(playerId?: string): Promise<AchievementProgress[]> {
    // If Tauri is not available, use fallback
    if (!invoke) {
      const defaultPlayerId = playerId || "default_user";
      return fallbackStorage.progress.get(defaultPlayerId) || [];
    }

    const responseJson = await safeInvoke<string>("get_player_progress", {
      playerId,
    });
    return JSON.parse(responseJson);
  }

  /**
   * Update achievement progress
   */
  async updateProgress(
    achievementId: string,
    progress: number,
    playerId?: string
  ): Promise<AchievementProgress> {
    const request: UpdateProgressRequest = {
      achievement_id: achievementId,
      progress,
      player_id: playerId,
    };

    // If Tauri is not available, use fallback
    if (!invoke) {
      const defaultPlayerId = playerId || "default_user";
      const now = Date.now();
      const progressRecord: AchievementProgress = {
        id: `${achievementId}_${defaultPlayerId}`,
        achievementId,
        playerId: defaultPlayerId,
        progress,
        maxProgress: 100,
        unlocked: progress >= 100,
        unlockedAt: progress >= 100 ? now : undefined,
        startedAt: now,
        lastUpdatedAt: now,
      };
      let playerProgress = fallbackStorage.progress.get(defaultPlayerId) || [];
      const index = playerProgress.findIndex((p) => p.achievementId === achievementId);
      if (index >= 0) {
        playerProgress[index] = progressRecord;
      } else {
        playerProgress.push(progressRecord);
      }
      fallbackStorage.progress.set(defaultPlayerId, playerProgress);
      return progressRecord;
    }

    const responseJson = await safeInvoke<string>("update_achievement_progress", {
      request,
    });
    return JSON.parse(responseJson);
  }

  /**
   * Unlock an achievement
   */
  async unlockAchievement(
    achievementId: string,
    context?: Record<string, unknown>,
    playerId?: string
  ): Promise<void> {
    const request: UnlockAchievementRequest = {
      achievement_id: achievementId,
      player_id: playerId,
      context_json: context ? JSON.stringify(context) : undefined,
    };

    // If Tauri is not available, use fallback
    if (!invoke) {
      const defaultPlayerId = playerId || "default_user";
      const unlock: AchievementUnlock = {
        id: `${achievementId}_${defaultPlayerId}_${Date.now()}`,
        achievementId,
        playerId: defaultPlayerId,
        unlockedAt: Date.now(),
        context: context,
      };
      let unlocks = fallbackStorage.unlocks.get(defaultPlayerId) || [];
      unlocks.push(unlock);
      fallbackStorage.unlocks.set(defaultPlayerId, unlocks);
      return;
    }

    await safeInvoke<string>("unlock_achievement", { request });
  }

  /**
   * Get player stats (level, XP, etc.)
   */
  async getPlayerStats(playerId?: string): Promise<PlayerStats> {
    // If Tauri is not available, use fallback
    if (!invoke) {
      const defaultPlayerId = playerId || "default_user";
      let stats = fallbackStorage.playerStats.get(defaultPlayerId);

      if (!stats) {
        stats = {
          player_id: defaultPlayerId,
          level: 1,
          xp: 0,
          total_xp_earned: 0,
          achievements_unlocked: 0,
          current_streak: 0,
          longest_streak: 0,
          created_at: Date.now(),
          updated_at: Date.now(),
        };
        fallbackStorage.playerStats.set(defaultPlayerId, stats);
      }

      return stats;
    }

    const responseJson = await safeInvoke<string>("get_player_stats", {
      playerId,
    });
    return JSON.parse(responseJson);
  }

  /**
   * Add XP to player
   */
  async addXp(amount: number, playerId?: string): Promise<PlayerStats> {
    const request: AddXpRequest = {
      amount,
      player_id: playerId,
    };

    // If Tauri is not available, use fallback
    if (!invoke) {
      const defaultPlayerId = playerId || "default_user";
      let stats = fallbackStorage.playerStats.get(defaultPlayerId);

      if (!stats) {
        stats = {
          player_id: defaultPlayerId,
          level: 1,
          xp: 0,
          total_xp_earned: 0,
          achievements_unlocked: 0,
          current_streak: 0,
          longest_streak: 0,
          created_at: Date.now(),
          updated_at: Date.now(),
        };
      }

      stats.xp += amount;
      stats.total_xp_earned += amount;
      stats.level = Math.floor(stats.total_xp_earned / 1000) + 1;
      stats.updated_at = Date.now();

      fallbackStorage.playerStats.set(defaultPlayerId, stats);
      return stats;
    }

    const responseJson = await safeInvoke<string>("add_xp", { request });
    return JSON.parse(responseJson);
  }

  /**
   * Get achievement analytics
   */
  async getAnalytics(playerId?: string): Promise<AchievementAnalytics> {
    // If Tauri is not available, use fallback
    if (!invoke) {
      const defaultPlayerId = playerId || "default_user";
      const unlocks = fallbackStorage.unlocks.get(defaultPlayerId) || [];
      const totalCount = fallbackStorage.achievements.size;
      const unlockedCount = unlocks.length;

      const rarityDistribution: Record<AchievementRarity, number> = {
        common: 0,
        rare: 0,
        epic: 0,
        legendary: 0,
      };

      const categoryDistribution: Record<AchievementCategory, number> = {
        tutorial: 0,
        creation: 0,
        collaboration: 0,
        mastery: 0,
        discovery: 0,
        social: 0,
        speed: 0,
        collection: 0,
        hidden: 0,
      };

      return {
        totalAchievements: totalCount,
        unlockedAchievements: unlockedCount,
        unlockPercentage: totalCount > 0 ? (unlockedCount / totalCount) * 100 : 0,
        rarityDistribution,
        rarityUnlockRates: { common: 0, rare: 0, epic: 0, legendary: 0 },
        categoryDistribution,
        categoryProgress: categoryDistribution,
        engagementScore: 0,
        averageTimeToUnlock: 0,
        recentUnlocks: [],
        currentStreak: 0,
        longestStreak: 0,
      };
    }

    const responseJson = await safeInvoke<string>("get_achievement_analytics", {
      playerId,
    });
    const data = JSON.parse(responseJson) as {
      totalCount: number;
      unlockedCount: number;
      unlockPercentage: number;
      rarityDistribution: Record<AchievementRarity, number>;
    };

    // Transform to match frontend type
    const rarityUnlockRates: Record<AchievementRarity, number> = {
      common: 0,
      rare: 0,
      epic: 0,
      legendary: 0,
    };

    const categoryDistribution: Record<AchievementCategory, number> = {
      tutorial: 0,
      creation: 0,
      collaboration: 0,
      mastery: 0,
      discovery: 0,
      social: 0,
      speed: 0,
      collection: 0,
      hidden: 0,
    };

    const categoryProgress: Record<AchievementCategory, number> = {
      tutorial: 0,
      creation: 0,
      collaboration: 0,
      mastery: 0,
      discovery: 0,
      social: 0,
      speed: 0,
      collection: 0,
      hidden: 0,
    };

    return {
      totalAchievements: data.totalCount,
      unlockedAchievements: data.unlockedCount,
      unlockPercentage: data.unlockPercentage,
      rarityDistribution: data.rarityDistribution,
      rarityUnlockRates,
      categoryDistribution,
      categoryProgress,
      engagementScore: 0,
      averageTimeToUnlock: 0,
      recentUnlocks: [],
      currentStreak: 0,
      longestStreak: 0,
    };
  }

  /**
   * Get unlock history
   */
  async getUnlockHistory(playerId?: string): Promise<AchievementUnlock[]> {
    // If Tauri is not available, use fallback
    if (!invoke) {
      const defaultPlayerId = playerId || "default_user";
      return fallbackStorage.unlocks.get(defaultPlayerId) || [];
    }

    const responseJson = await safeInvoke<string>("get_unlock_history", {
      playerId,
    });
    return JSON.parse(responseJson);
  }

  /**
   * Get user progress (combines stats and achievements)
   */
  async getUserProgress(playerId?: string): Promise<UserProgress> {
    const stats = await this.getPlayerStats(playerId);
    const unlockedAchievements = await this.getUnlockHistory(playerId);
    const progress = playerStatsToUserProgress(stats);
    progress.achievements = unlockedAchievements.map((u) => u.achievementId || "");
    return progress;
  }

  /**
   * Complete a mission (for store compatibility)
   */
  async completeMission(playerId: string): Promise<UserProgress> {
    // Award XP for mission completion
    const newStats = await this.addXp(1000, playerId); // Default 1000 XP per mission
    const progress = playerStatsToUserProgress(newStats);
    const unlockedAchievements = await this.getUnlockHistory(playerId);
    progress.achievements = unlockedAchievements.map((u) => u.achievementId || "");
    return progress;
  }

  /**
   * Get available achievements (not yet unlocked)
   */
  getAvailableAchievements(allAchievements: Achievement[], progress: UserProgress): Achievement[] {
    return allAchievements.filter((a) => !progress.achievements.includes(a.id) && !a.hidden);
  }

  /**
   * Get unlocked achievements
   */
  getUnlockedAchievements(allAchievements: Achievement[], progress: UserProgress): Achievement[] {
    return allAchievements.filter((a) => progress.achievements.includes(a.id));
  }

  /**
   * Get locked achievements
   */
  getLockedAchievements(allAchievements: Achievement[], progress: UserProgress): Achievement[] {
    return allAchievements.filter((a) => !progress.achievements.includes(a.id) && a.hidden);
  }
}

// Singleton instance
export const achievementDB = new AchievementDatabaseAPI();

// Export the class for type inference
export const achievementAPI = achievementDB;

// Extended player progress type for UI consumption
export interface UserProgress {
  playerId: string;
  level: number;
  xp: number;
  currentXp: number;
  achievements: string[];
  completedMissions: string[];
  totalXpEarned: number;
  achievementsUnlocked: number;
  currentStreak: number;
  longestStreak: number;
}

export interface LevelInfo {
  level: number;
  currentXp: number;
  nextLevelXp: number;
  progressPercentage: number;
}

// Helper function to convert PlayerStats to UserProgress
export function playerStatsToUserProgress(stats: PlayerStats): UserProgress {
  return {
    playerId: stats.player_id,
    level: stats.level,
    xp: stats.xp,
    currentXp: stats.xp % 1000, // XP in current level
    achievements: [], // Will be populated by store
    completedMissions: [], // Will be populated by store
    totalXpEarned: stats.total_xp_earned,
    achievementsUnlocked: stats.achievements_unlocked,
    currentStreak: stats.current_streak,
    longestStreak: stats.longest_streak,
  };
}

// Helper function to calculate level info
export function calculateLevelInfo(xp: number): LevelInfo {
  const level = Math.floor(xp / 1000) + 1;
  const currentXp = xp % 1000;
  const nextLevelXp = 1000;
  const progressPercentage = (currentXp / nextLevelXp) * 100;

  return {
    level,
    currentXp,
    nextLevelXp,
    progressPercentage,
  };
}

File: src-tauri/src/commands/achievements.rs
/**
 * Tauri Commands for Achievement System
 * Exposes achievement database operations to the frontend
 */
use crate::db::{
    Achievement, AchievementDb, AchievementProgress, AchievementUnlock, Database, PlayerStats,
};
use serde::{Deserialize, Serialize};
use std::sync::Mutex;
use tauri::State;

#[derive(Debug, Serialize, Deserialize)]
pub struct InitAchievementsRequest {
    pub achievements_json: String, // Serialized array of achievements from frontend
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UpdateProgressRequest {
    pub achievement_id: String,
    pub progress: i64,
    pub player_id: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UnlockAchievementRequest {
    pub achievement_id: String,
    pub player_id: Option<String>,
    pub context_json: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AddXpRequest {
    pub amount: i64,
    pub player_id: Option<String>,
}

// ============================================
// INITIALIZATION
// ============================================

#[tauri::command]
pub async fn init_achievements(
    db: State<'_, Mutex<Database>>,
    request: InitAchievementsRequest,
) -> Result<String, String> {
    let db = db.lock().map_err(|e| e.to_string())?;
    let conn = db.get_connection();
    let achievement_db = AchievementDb::new(conn);

    // Parse achievements from JSON
    let achievements: Vec<Achievement> = serde_json::from_str(&request.achievements_json)
        .map_err(|e| format!("Failed to parse achievements: {}", e))?;

    // Insert all achievements
    let count = achievements.len();
    for achievement in achievements {
        achievement_db
            .insert_achievement(&achievement)
            .map_err(|e| format!("Failed to insert achievement: {}", e))?;
    }

    Ok(format!("Initialized {} achievements", count))
}

// ============================================
// ACHIEVEMENT QUERIES
// ============================================

#[tauri::command]
pub async fn get_all_achievements(db: State<'_, Mutex<Database>>) -> Result<String, String> {
    let db = db.lock().map_err(|e| e.to_string())?;
    let conn = db.get_connection();
    let achievement_db = AchievementDb::new(conn);

    let achievements = achievement_db
        .get_all_achievements()
        .map_err(|e| format!("Failed to get achievements: {}", e))?;

    serde_json::to_string(&achievements)
        .map_err(|e| format!("Failed to serialize achievements: {}", e))
}

#[tauri::command]
pub async fn get_achievement_by_id(
    db: State<'_, Mutex<Database>>,
    achievement_id: String,
) -> Result<String, String> {
    let db = db.lock().map_err(|e| e.to_string())?;
    let conn = db.get_connection();
    let achievement_db = AchievementDb::new(conn);

    let achievement = achievement_db
        .get_achievement_by_id(&achievement_id)
        .map_err(|e| format!("Failed to get achievement: {}", e))?;

    serde_json::to_string(&achievement)
        .map_err(|e| format!("Failed to serialize achievement: {}", e))
}

// ============================================
// PROGRESS TRACKING
// ============================================

#[tauri::command]
pub async fn get_player_progress(
    db: State<'_, Mutex<Database>>,
    player_id: Option<String>,
) -> Result<String, String> {
    let player_id = player_id.unwrap_or_else(|| "local-user".to_string());

    let db = db.lock().map_err(|e| e.to_string())?;
    let conn = db.get_connection();
    let achievement_db = AchievementDb::new(conn);

    let progress = achievement_db
        .get_progress_by_player(&player_id)
        .map_err(|e| format!("Failed to get progress: {}", e))?;

    serde_json::to_string(&progress).map_err(|e| format!("Failed to serialize progress: {}", e))
}

#[tauri::command]
pub async fn update_achievement_progress(
    db: State<'_, Mutex<Database>>,
    request: UpdateProgressRequest,
) -> Result<String, String> {
    let player_id = request
        .player_id
        .unwrap_or_else(|| "local-user".to_string());
    let now = crate::db::achievements::get_current_timestamp();

    let db = db.lock().map_err(|e| e.to_string())?;
    let conn = db.get_connection();
    let achievement_db = AchievementDb::new(conn);

    // Get or create progress record
    let mut progress = match achievement_db
        .get_progress(&request.achievement_id, &player_id)
        .map_err(|e| format!("Failed to get progress: {}", e))?
    {
        Some(p) => p,
        None => {
            // Get achievement to determine max_progress
            let achievement = achievement_db
                .get_achievement_by_id(&request.achievement_id)
                .map_err(|e| format!("Failed to get achievement: {}", e))?
                .ok_or_else(|| format!("Achievement not found: {}", request.achievement_id))?;

            // Parse condition to get max_progress
            let condition: serde_json::Value = serde_json::from_str(&achievement.condition_data)
                .map_err(|e| format!("Failed to parse condition: {}", e))?;

            let max_progress = match condition.get("target") {
                Some(target) => target.as_i64().unwrap_or(1),
                None => 1,
            };

            AchievementProgress {
                id: crate::db::achievements::generate_id(),
                achievement_id: request.achievement_id.clone(),
                player_id: player_id.clone(),
                progress: 0,
                max_progress,
                unlocked: false,
                unlocked_at: None,
                started_at: now,
                last_updated_at: now,
                metadata: None,
            }
        }
    };

    // Update progress
    progress.progress = request.progress;
    progress.last_updated_at = now;

    // Check if unlocked
    if progress.progress >= progress.max_progress && !progress.unlocked {
        progress.unlocked = true;
        progress.unlocked_at = Some(now);
    }

    // Save to database
    achievement_db
        .upsert_progress(&progress)
        .map_err(|e| format!("Failed to update progress: {}", e))?;

    serde_json::to_string(&progress).map_err(|e| format!("Failed to serialize progress: {}", e))
}

// ============================================
// UNLOCKING
// ============================================

#[tauri::command]
pub async fn unlock_achievement(
    db: State<'_, Mutex<Database>>,
    request: UnlockAchievementRequest,
) -> Result<String, String> {
    let player_id = request
        .player_id
        .unwrap_or_else(|| "local-user".to_string());
    let now = crate::db::achievements::get_current_timestamp();

    let db = db.lock().map_err(|e| e.to_string())?;
    let conn = db.get_connection();
    let achievement_db = AchievementDb::new(conn);

    // Record unlock
    let unlock = AchievementUnlock {
        id: crate::db::achievements::generate_id(),
        player_id: player_id.clone(),
        achievement_id: request.achievement_id.clone(),
        unlocked_at: now,
        context: request.context_json,
    };

    achievement_db
        .record_unlock(&unlock)
        .map_err(|e| format!("Failed to record unlock: {}", e))?;

    // Update progress to unlocked
    let mut progress = achievement_db
        .get_progress(&request.achievement_id, &player_id)
        .map_err(|e| format!("Failed to get progress: {}", e))?
        .ok_or_else(|| "Progress record not found".to_string())?;

    progress.unlocked = true;
    progress.unlocked_at = Some(now);
    progress.last_updated_at = now;

    achievement_db
        .upsert_progress(&progress)
        .map_err(|e| format!("Failed to update progress: {}", e))?;

    Ok("Achievement unlocked".to_string())
}

// ============================================
// PLAYER STATS
// ============================================

#[tauri::command]
pub async fn get_player_stats(
    db: State<'_, Mutex<Database>>,
    player_id: Option<String>,
) -> Result<String, String> {
    let player_id = player_id.unwrap_or_else(|| "local-user".to_string());

    let db = db.lock().map_err(|e| e.to_string())?;
    let conn = db.get_connection();
    let achievement_db = AchievementDb::new(conn);

    let stats = achievement_db
        .get_player_stats(&player_id)
        .map_err(|e| format!("Failed to get player stats: {}", e))?
        .ok_or_else(|| "Player stats not found".to_string())?;

    serde_json::to_string(&stats).map_err(|e| format!("Failed to serialize stats: {}", e))
}

#[tauri::command]
pub async fn add_xp(
    db: State<'_, Mutex<Database>>,
    request: AddXpRequest,
) -> Result<String, String> {
    let player_id = request
        .player_id
        .unwrap_or_else(|| "local-user".to_string());
    let now = crate::db::achievements::get_current_timestamp();

    let db = db.lock().map_err(|e| e.to_string())?;
    let conn = db.get_connection();
    let achievement_db = AchievementDb::new(conn);

    // Get or create player stats
    let mut stats = match achievement_db
        .get_player_stats(&player_id)
        .map_err(|e| format!("Failed to get player stats: {}", e))?
    {
        Some(stats) => stats,
        None => {
            // Create new player stats if not found
            let new_stats = PlayerStats {
                player_id: player_id.clone(),
                level: 1,
                xp: 0,
                total_xp_earned: 0,
                achievements_unlocked: 0,
                current_streak: 0,
                longest_streak: 0,
                last_login: Some(now),
                created_at: now,
                updated_at: now,
            };

            achievement_db
                .update_player_stats(&new_stats)
                .map_err(|e| format!("Failed to create player stats: {}", e))?;

            new_stats
        }
    };

    // Add XP
    stats.xp += request.amount;
    stats.total_xp_earned += request.amount;

    // Calculate level (using exponential curve: level = floor(1.5^(level-1) * 100))
    let mut new_level = stats.level;
    loop {
        let xp_for_next_level = (100.0 * (1.5_f64).powi((new_level) as i32)) as i64;
        if stats.xp >= xp_for_next_level {
            stats.xp -= xp_for_next_level;
            new_level += 1;
        } else {
            break;
        }
    }
    stats.level = new_level;
    stats.updated_at = now;

    achievement_db
        .update_player_stats(&stats)
        .map_err(|e| format!("Failed to update stats: {}", e))?;

    serde_json::to_string(&stats).map_err(|e| format!("Failed to serialize stats: {}", e))
}

// ============================================
// ANALYTICS
// ============================================

#[tauri::command]
pub async fn get_achievement_analytics(
    db: State<'_, Mutex<Database>>,
    player_id: Option<String>,
) -> Result<String, String> {
    let player_id = player_id.unwrap_or_else(|| "local-user".to_string());

    let db = db.lock().map_err(|e| e.to_string())?;
    let conn = db.get_connection();
    let achievement_db = AchievementDb::new(conn);

    let unlocked_count = achievement_db
        .get_unlock_count(&player_id)
        .map_err(|e| format!("Failed to get unlock count: {}", e))?;

    let total_count = achievement_db
        .get_total_achievements()
        .map_err(|e| format!("Failed to get total achievements: {}", e))?;

    let rarity_dist = achievement_db
        .get_rarity_distribution(&player_id)
        .map_err(|e| format!("Failed to get rarity distribution: {}", e))?;

    let analytics = serde_json::json!({
        "unlockedCount": unlocked_count,
        "totalCount": total_count,
        "unlockPercentage": if total_count > 0 {
            (unlocked_count as f64 / total_count as f64) * 100.0
        } else {
            0.0
        },
        "rarityDistribution": rarity_dist.into_iter().collect::<std::collections::HashMap<_, _>>(),
    });

    serde_json::to_string(&analytics).map_err(|e| format!("Failed to serialize analytics: {}", e))
}

#[tauri::command]
pub async fn get_unlock_history(
    db: State<'_, Mutex<Database>>,
    player_id: Option<String>,
) -> Result<String, String> {
    let player_id = player_id.unwrap_or_else(|| "local-user".to_string());

    let db = db.lock().map_err(|e| e.to_string())?;
    let conn = db.get_connection();
    let achievement_db = AchievementDb::new(conn);

    let unlocks = achievement_db
        .get_unlocks_by_player(&player_id)
        .map_err(|e| format!("Failed to get unlocks: {}", e))?;

    serde_json::to_string(&unlocks).map_err(|e| format!("Failed to serialize unlocks: {}", e))
}
